spec/puppet-server-lib/puppet/jvm/ast_compiler_spec.rb

```diff
@@ -0,0 +1,58 @@
+require 'spec_helper'
+
+require 'puppet_pal'
+
+require 'puppet/server/ast_compiler'
+
+def generate_ast(code)
+  Puppet::Pal.in_tmp_environment("ast") do |pal|
+    pal.with_catalog_compiler do |compiler|
+      ast = compiler.parse_string(code)
+      Puppet::Pops::Serialization::ToDataConverter.convert(ast,
+                                                           rich_data: true,
+                                                           symbol_to_string: true)
+    end
+  end
+end
+
+def request(ast)
+  {"certname" => "localhost",
+   "facts" => {"values" => {"my_fact" => "fact_value"}},
+   "trusted_facts" => {"values" => {"my_trusted" => "trusted_value"}},
+   "options" => {"capture_logs" => false},
+   "environment" => "production",
+   "code_ast" => ast.to_json,
+   "variables" => {"values" => {"foo" => "bar"}}}
+end
+
+describe Puppet::Server::ASTCompiler do
+  context 'when compiling AST' do
+    it 'handles basic resources' do
+      ast_notify = generate_ast("notify { 'maggie': }")
+      req = request(ast_notify)
+      response = Puppet::Server::ASTCompiler.compile(req)
+      expect(response[:catalog]).to match(/Notify\[maggie\]/)
+    end
+
+    it 'correctly interpolates supplied variables' do
+      ast_variable = generate_ast('notify { "$foo": }')
+      req = request(ast_variable)
+      response = Puppet::Server::ASTCompiler.compile(req)
+      expect(response[:catalog]).to match(/Notify\[bar\]/)
+    end
+
+    it 'correctly interpolates supplied facts' do
+      ast_facts = generate_ast('notify { "$my_fact": }')
+      req = request(ast_facts)
+      response = Puppet::Server::ASTCompiler.compile(req)
+      expect(response[:catalog]).to match(/Notify\[fact_value\]/)
+    end
+
+    it 'correctly interpolates supplied trusted facts' do
+      ast_trusted_facts = generate_ast('notify { "${trusted[\'my_trusted\']}": }')
+      req = request(ast_trusted_facts)
+      response = Puppet::Server::ASTCompiler.compile(req)
+      expect(response[:catalog]).to match(/Notify\[trusted_value\]/)
```

Comment:
Does the `compile` method return a Hash with a `:catalog` key that is JSON string? Can we just return a Hash and do serialization all at once?

Reply:
I don't think so, because of the interface of the PAL catalog compiler. The only way to get the catalog out is via the `with_json_encoding` method. The other methods don't actually return the catalog. So we'd have to deserialize that only to re-serialize it later.

Reply:
Do you think it makes a difference? We could probably add something to the API if we really think it's better to do it all at once.

Reply:
it feels like a weird way to shape the response internally and the reason seems to be because the thing we're calling into has a weird interface. do you know why the interface to compiling in pal is that way?

aside, in our other endpoint we get a data structure that represents the catalog (from calling `to_data_hash` on a Catalog instance) and then serialize the whole thing at once. that feels more natural but if there's a good reason that pal needs to do the serializing then maybe there's a latent bug in the other endpoint?

Reply:
So, there is some post-processing that PAL does as part of the JSON encoding. It gives you the option to exclude virtual resources (this is on by default) and will clean up any resources that are missing names or titles. 

That being said, by simply making the [`catalog` method](https://github.com/puppetlabs/puppet/blob/master/lib/puppet/pal/catalog_compiler.rb#L12-L16) public, we have access to the same catalog object that we turn to a hash in the other endpoint. I dug through the history, and that method has been private since the beginning, with no comment as to why. But I don't think it would be too bad to expose it... This seems like it was maybe just designed for another use case (a human use case seems probable, given that pretty printing is also on by default).

Reply:
As mentioned in the other PR where an API was added - the reason is that we don't want all of the `Catalog` implementation to be considered API - and it should instead be available via a wrapper that provides the parts we do want as API. In the other PR, a method is added to serialize the catalog as a data_hash - that is fine as API.

Reply:
I'll update this now that the other PR is merged and the API finalized.